loopin.go

```diff
@@ -0,0 +1,585 @@
+
+	// If an unexpected error happened, report a temporary failure
+	// but don't persist the error. Otherwise for example a
+	// connection error could lead to abandoning the swap
+	// permanently and losing funds.
+	if err != nil {
+		s.log.Errorf("Swap error: %v", err)
+		s.state = loopdb.StateFailTemporary
+
+		// If we cannot send out this update, there is nothing we can do.
+		_ = s.sendUpdate(mainCtx)
+
+		return err
+	}
+
+	s.log.Infof("Loop in swap completed: %v "+
+		"(final cost: server %v, onchain %v)",
+		s.state,
+		s.cost.Server,
+		s.cost.Onchain,
+	)
+
+	return nil
+}
+
+// executeSwap executes the swap.
+func (s *loopInSwap) executeSwap(globalCtx context.Context) error {
+	var err error
+
+	// For loop in, the client takes the first step by publishing the
+	// on-chain htlc. Only do this is we haven't already done so in a
+	// previous run.
+	if s.state == loopdb.StateInitiated {
+		published, err := s.publishOnChainHtlc(globalCtx)
+		if err != nil {
+			return err
+		}
+		if !published {
+			return nil
+		}
+	}
+
+	// Wait for the htlc to confirm. After a restart this will pick up a
+	// previously published tx.
+	conf, err := s.waitForHtlcConf(globalCtx)
+	if err != nil {
+		return err
+	}
+
+	// Determine the htlc outpoint by inspecting the htlc tx.
+	htlcOutpoint, htlcValue, err := swap.GetScriptOutput(
+		conf.Tx, s.htlc.ScriptHash,
+	)
+	if err != nil {
+		return err
+	}
+
+	// TODO: Add miner fee of htlc tx to swap cost balance.
+
+	// The server is expected to see the htlc on-chain and knowing that it
+	// can sweep that htlc with the preimage, it should pay our swap
+	// invoice, receive the preimage and sweep the htlc. We are waiting for
+	// this to happen and simultaneously watch the htlc expiry height. When
+	// the htlc expires, we will publish a timeout tx to reclaim the funds.
+	spend, err := s.waitForHtlcSpend(globalCtx, htlcOutpoint)
+	if err != nil {
+		return err
+	}
+
+	// Determine the htlc input of the spending tx and inspect the witness
+	// to findout whether a success or a timeout tx spend the htlc.
+	htlcInput := spend.SpendingTx.TxIn[spend.SpenderInputIndex]
+
+	if s.htlc.IsSuccessWitness(htlcInput.Witness) {
+		s.state = loopdb.StateSuccess
+
+		// Server swept the htlc. The htlc value can be added to the
+		// server cost balance.
+		s.cost.Server += htlcValue
+	} else {
+		s.state = loopdb.StateFailTimeout
+
+		// Now that the timeout tx confirmed, we can safely cancel the
+		// swap invoice. We still need to query the final invoice state.
+		// This is not a hodl invoice, so it may be that the invoice was
+		// already settled. This means that the server didn't succeed in
+		// sweeping the htlc after paying the invoice.
+		err := s.lnd.Invoices.CancelInvoice(globalCtx, s.hash)
+		if err != nil && err != channeldb.ErrInvoiceAlreadySettled {
+			return err
+		}
+
+		// TODO: Add miner fee of timeout tx to swap cost balance.
+	}
+
+	// Wait for a final state of the swap invoice. It should either be
```

Comment:
I think we should move this to be _above_ waiting for the HTLC spend, or be concurrent even. I tried another Loop in on testnet now and realized that things are shown as pending until the HTLC is swept on chain. However, from the PoV of the client, the Loop In process is complete once its invoice is settled, as it now has more outbound capacity.

Reply:
Everywhere in loop swaps are only marked complete when all information needed for accounting is there. For loop in on the client, this means the htlc sweep tx needs to be confirmed and the swap invoice either canceled or settled. Also after a restart, a swap that is completed from the user PoV, still needs to be resumed to complete the accounting info.

An intermediate state that the swap transitions to when the invoice is paid could be a solution here.

Reply:
Added a commit that adds this intermediate state.